### PR TITLE
Add search based physical execution steps

### PR DESCRIPTION
**Pull request by szdavid92**
- [x] Expressions for property checks
- [x] Basic iterators (vertex/edge)
- [ ] Iterator combinators for trees (multiple HEADS)
- [ ] Find out what else is needed

### CODE REVIEW

File: ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Core.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package sre.task
+
+object Core {
+  import ingraph.ire.{IngraphEdge, IngraphElement, IngraphObject, IngraphVertex}
+  type ElementT[A] = IngraphElement[A]
```
Comment (szdavid92) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Core.scala]: remove

File: ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    def check(assert: BExpr[VertexT]): Iterator[VertexT] =
+      new VertexIterator(iterator.filter((vertex) => assert(vertex)))
+
+    def extend(direction: Direction.Value, label: Option[String]): EdgeIterator = {
+      val iteratorFromMapVariant = label match {
+        case Some(label) => (x: scala.collection.Map[String, EdgeT])
+          => x.get(label).iterator // we can create an iterator out of an Option
+        case _ => (x: scala.collection.Map[String, EdgeT])
+          // FIXME: if we don't have any label criteria, we should get all edges, right?
```
Comment (szdavid92) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala]: @szarnyasg

File: ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          => x.valuesIterator
+      }
+      val directionGetterVariant = direction match {
+        case Direction.Out => (v: VertexT, iteratorFromMap: (scala.collection.Map[String, EdgeT] => Iterator[EdgeT]))
+          => iteratorFromMap(v.edges)
+        case Direction.In => (v: VertexT, iteratorFromMap: (scala.collection.Map[String, EdgeT] => Iterator[EdgeT]))
+          => iteratorFromMap(v.reverseEdges.asInstanceOf[scala.collection.Map[String, EdgeT]])
```
Comment (szdavid92) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala]: remove redundant cast

File: ingraph-engine/ingraph-engine-sre/src/main/scala/util/TypeGadgets.scala [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package util
```
Comment (szdavid92) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/util/TypeGadgets.scala]: remove dead code

File: ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala (hunk 2 above)
Comment (szarnyasg) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Task.scala]: Yes, the comment is correct

File: ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Expressions.scala [hunk 5]
```diff
@@ -1,67 +1,182 @@
...
+        for { a <- optA; b <- optB } yield
+          strictOp(a, b)
+      }
+    }
   }
-  final case class Gte[OrdT : Ordering](rhs: OrdT) extends BExpr[Any] {
-    def apply(lhs: Any): Boolean =
-      implicitly[Ordering[OrdT]].gteq(lhs.asInstanceOf[OrdT], rhs) // Type should already be asserted
+
+  /* FIXME: I am not sure if all types are covered here. ask @szarnyasg
```
Comment (szdavid92) [inline ingraph-engine/ingraph-engine-sre/src/main/scala/sre/task/Expressions.scala]: @szarnyasg

**Pull request closed without merging**